File: app/src/main/java/com/example/doan/Activities/OtpActivity.java
package com.example.doan.Activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.doan.Models.ApiResponse;
import com.example.doan.Models.RegisterRequest;
import com.example.doan.Models.VerifyOtpRequest;
import com.example.doan.Network.RetrofitClient;
import com.example.doan.R;

import org.json.JSONObject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class OtpActivity extends AppCompatActivity {

    private static final String TAG = "OtpActivity";

    private EditText otpInput;
    private Button verifyButton;
    private TextView resendOtpText;

    private String username, password, email;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_otp);

        otpInput = findViewById(R.id.input_otp);
        verifyButton = findViewById(R.id.btn_verify_otp);
        resendOtpText = findViewById(R.id.text_resend_otp);

        // Nhận dữ liệu từ RegisterActivity
        Intent intent = getIntent();
        if (intent.hasExtra("USERNAME") && intent.hasExtra("PASSWORD") && intent.hasExtra("EMAIL")) {
            username = intent.getStringExtra("USERNAME");
            password = intent.getStringExtra("PASSWORD");
            email = intent.getStringExtra("EMAIL");

            // Gọi API đăng ký để server gửi OTP
            requestRegistrationAndOtp();
        } else {
            Toast.makeText(this, "Lỗi: Không tìm thấy thông tin đăng ký.", Toast.LENGTH_LONG).show();
            finish(); // Quay lại trang trước
            return;
        }

        verifyButton.setOnClickListener(v -> {
            String otp = otpInput.getText().toString().trim();
            if (otp.length() != 6) {
                Toast.makeText(this, "Mã OTP phải có 6 chữ số.", Toast.LENGTH_SHORT).show();
                return;
            }
            verifyOtp(otp);
        });

        resendOtpText.setOnClickListener(v -> resendOtp());
    }

    private void requestRegistrationAndOtp() {
        Toast.makeText(this, "Đang gửi yêu cầu đăng ký...", Toast.LENGTH_SHORT).show();
        RegisterRequest registerRequest = new RegisterRequest(username, email, password, username, "");

        RetrofitClient.getInstance(this).getApiService().registerWithOtp(registerRequest)
                .enqueue(new Callback<ApiResponse<String>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<String>> call, @NonNull Response<ApiResponse<String>> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().isSuccess()) {
                            Toast.makeText(OtpActivity.this, response.body().getMessage(), Toast.LENGTH_SHORT).show();
                        } else {
                            String errorMessage = "Lỗi khi gửi OTP.";
                            try {
                                if (response.body() != null) {
                                    errorMessage = response.body().getMessage();
                                } else if (response.errorBody() != null) {
                                    errorMessage = new JSONObject(response.errorBody().string()).getString("message");
                                }
                            } catch (Exception e) {
                                Log.e(TAG, "Error parsing error response in requestRegistrationAndOtp", e);
                            }
                            Toast.makeText(OtpActivity.this, errorMessage, Toast.LENGTH_LONG).show();
                            // Cân nhắc finish() để quay về trang đăng ký nếu có lỗi
                            // finish();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<String>> call, @NonNull Throwable t) {
                        Log.e(TAG, "Lỗi mạng khi đăng ký: " + t.getMessage());
//                        Toast.makeText(OtpActivity.this, "Lỗi mạng, không thể kết nối server.", Toast.LENGTH_SHORT).show();
                        // Cân nhắc finish() để quay về trang đăng ký nếu có lỗi
                        // finish();
                    }
                });
    }

    private void verifyOtp(String otp) {
        VerifyOtpRequest request = new VerifyOtpRequest(email, otp);

        RetrofitClient.getInstance(this).getApiService().verifyOtp(request)
                .enqueue(new Callback<ApiResponse<String>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<String>> call, @NonNull Response<ApiResponse<String>> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().isSuccess()) {
                            Toast.makeText(OtpActivity.this, "Kích hoạt thành công! Vui lòng đăng nhập.", Toast.LENGTH_LONG).show();
                            Intent intent = new Intent(OtpActivity.this, LoginActivity.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                            startActivity(intent);
                            finish();
                        } else {
                            String errorMessage = "Lỗi xác thực.";
                            try {
                                if (response.errorBody() != null) {
                                    String errorBodyStr = response.errorBody().string();
                                    JSONObject errorObj = new JSONObject(errorBodyStr);
                                    if (errorObj.has("message")) {
                                        errorMessage = errorObj.getString("message");
                                    }
                                }
                            } catch (Exception e) {
                                Log.e(TAG, "Error parsing error body in verifyOtp", e);
                            }
                            Toast.makeText(OtpActivity.this, errorMessage, Toast.LENGTH_LONG).show();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<String>> call, @NonNull Throwable t) {
                        Log.e(TAG, "Lỗi Retrofit khi xác thực OTP: " + t.getMessage());
                        Toast.makeText(OtpActivity.this, "Lỗi kết nối: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void resendOtp() {
        RetrofitClient.getInstance(this).getApiService().resendOtp(email)
                .enqueue(new Callback<ApiResponse<String>>() {
                    @Override
                    public void onResponse(@NonNull Call<ApiResponse<String>> call, @NonNull Response<ApiResponse<String>> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().isSuccess()) {
                            Toast.makeText(OtpActivity.this, "OTP đã được gửi lại!", Toast.LENGTH_SHORT).show();
                        } else {
                            String errorMessage = "Lỗi gửi lại OTP.";
                            if(response.body() != null) {
                                errorMessage = response.body().getMessage();
                            } else if (response.errorBody() != null) {
                                try {
                                    errorMessage = new JSONObject(response.errorBody().string()).getString("message");
                                } catch(Exception e) { /* Do nothing */ }
                            }
                            Toast.makeText(OtpActivity.this, errorMessage, Toast.LENGTH_LONG).show();
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<ApiResponse<String>> call, @NonNull Throwable t) {
                        Log.e(TAG, "Network Error on resend: " + t.getMessage());
                        Toast.makeText(OtpActivity.this, "Không thể kết nối đến server.", Toast.LENGTH_SHORT).show();
                    }
                });
    }
}

File: app/src/main/java/com/example/doan/Activities/OtpActivity.kt
package com.example.doan.Activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.doan.Models.ApiResponse
import com.example.doan.Models.RegisterRequest
import com.example.doan.Models.VerifyOtpRequest
import com.example.doan.Network.RetrofitClient
import com.example.doan.R
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class OtpActivity : AppCompatActivity() {

    private lateinit var otpInput: EditText
    private lateinit var verifyButton: Button
    private lateinit var resendOtpText: TextView

    private var username: String = ""
    private var password: String = ""
    private var email: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_otp)

        otpInput = findViewById(R.id.input_otp)
        verifyButton = findViewById(R.id.btn_verify_otp)
        resendOtpText = findViewById(R.id.text_resend_otp)

        // Nhận dữ liệu từ RegisterActivity
        intent?.let {
            if (it.hasExtra("USERNAME") && it.hasExtra("PASSWORD") && it.hasExtra("EMAIL")) {
                username = it.getStringExtra("USERNAME") ?: ""
                password = it.getStringExtra("PASSWORD") ?: ""
                email = it.getStringExtra("EMAIL") ?: ""

                // Gọi API đăng ký để server gửi OTP
                requestRegistrationAndOtp()
            } else {
                Toast.makeText(this, "Lỗi: Không tìm thấy thông tin đăng ký.", Toast.LENGTH_LONG).show()
                finish()
                return
            }
        }

        verifyButton.setOnClickListener {
            val otp = otpInput.text.toString().trim()
            if (otp.length != 6) {
                Toast.makeText(this, "Mã OTP phải có 6 chữ số.", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            verifyOtp(otp)
        }

        resendOtpText.setOnClickListener { resendOtp() }
    }

    private fun requestRegistrationAndOtp() {
        Toast.makeText(this, "Đang gửi yêu cầu đăng ký...", Toast.LENGTH_SHORT).show()
        val registerRequest = RegisterRequest(username, email, password, username, "")

        RetrofitClient.getInstance(this).apiService.registerWithOtp(registerRequest)
            .enqueue(object : Callback<ApiResponse<String>> {
                override fun onResponse(call: Call<ApiResponse<String>>, response: Response<ApiResponse<String>>) {
                    if (response.isSuccessful && response.body()?.success == true) {
                        Toast.makeText(this@OtpActivity, response.body()?.message, Toast.LENGTH_SHORT).show()
                    } else {
                        var errorMessage = "Lỗi khi gửi OTP."
                        try {
                            errorMessage = response.body()?.message 
                                ?: response.errorBody()?.string()?.let { JSONObject(it).getString("message") }
                                ?: errorMessage
                        } catch (e: Exception) {
                            Log.e(TAG, "Error parsing error response in requestRegistrationAndOtp", e)
                        }
                        Toast.makeText(this@OtpActivity, errorMessage, Toast.LENGTH_LONG).show()
                    }
                }

                override fun onFailure(call: Call<ApiResponse<String>>, t: Throwable) {
                    Log.e(TAG, "Lỗi mạng khi đăng ký: ${t.message}")
                }
            })
    }

    private fun verifyOtp(otp: String) {
        val request = VerifyOtpRequest(email, otp)

        RetrofitClient.getInstance(this).apiService.verifyOtp(request)
            .enqueue(object : Callback<ApiResponse<String>> {
                override fun onResponse(call: Call<ApiResponse<String>>, response: Response<ApiResponse<String>>) {
                    if (response.isSuccessful && response.body()?.success == true) {
                        Toast.makeText(this@OtpActivity, "Kích hoạt thành công! Vui lòng đăng nhập.", Toast.LENGTH_LONG).show()
                        val intent = Intent(this@OtpActivity, LoginActivity::class.java).apply {
                            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                        }
                        startActivity(intent)
                        finish()
                    } else {
                        var errorMessage = "Lỗi xác thực."
                        try {
                            response.errorBody()?.string()?.let { errorBodyStr ->
                                val errorObj = JSONObject(errorBodyStr)
                                if (errorObj.has("message")) {
                                    errorMessage = errorObj.getString("message")
                                }
                            }
                        } catch (e: Exception) {
                            Log.e(TAG, "Error parsing error body in verifyOtp", e)
                        }
                        Toast.makeText(this@OtpActivity, errorMessage, Toast.LENGTH_LONG).show()
                    }
                }

                override fun onFailure(call: Call<ApiResponse<String>>, t: Throwable) {
                    Log.e(TAG, "Lỗi Retrofit khi xác thực OTP: ${t.message}")
                    Toast.makeText(this@OtpActivity, "Lỗi kết nối: ${t.message}", Toast.LENGTH_SHORT).show()
                }
            })
    }

    private fun resendOtp() {
        RetrofitClient.getInstance(this).apiService.resendOtp(email)
            .enqueue(object : Callback<ApiResponse<String>> {
                override fun onResponse(call: Call<ApiResponse<String>>, response: Response<ApiResponse<String>>) {
                    if (response.isSuccessful && response.body()?.success == true) {
                        Toast.makeText(this@OtpActivity, "OTP đã được gửi lại!", Toast.LENGTH_SHORT).show()
                    } else {
                        var errorMessage = "Lỗi gửi lại OTP."
                        errorMessage = response.body()?.message 
                            ?: try {
                                response.errorBody()?.string()?.let { JSONObject(it).getString("message") }
                            } catch (e: Exception) { null }
                            ?: errorMessage
                        Toast.makeText(this@OtpActivity, errorMessage, Toast.LENGTH_LONG).show()
                    }
                }

                override fun onFailure(call: Call<ApiResponse<String>>, t: Throwable) {
                    Log.e(TAG, "Network Error on resend: ${t.message}")
                    Toast.makeText(this@OtpActivity, "Không thể kết nối đến server.", Toast.LENGTH_SHORT).show()
                }
            })
    }

    companion object {
        private const val TAG = "OtpActivity"
    }
}
